### ISSUE TITLE
Aero fighters attacking ground targets create deadzone during weapons phase?

### ISSUE THREAD

**Issue by VampireSeraphin** [MegaMek/megamek#332]
I'm not sure if this is a bug, or a cool little feature.

I was playing a game tonight where both myself, and my opponent had aerospace fighters in play on a ground map. He had me dead to rights, with two fighters directly behind mine, but his attacks vanished into the ether, hitting nothing. I don't think they were even displayed. 

During that same attack, rather than fire my rear laser, I choose to dive bomb his mech instead which I had over flown.

I think the reason his attacks disappeared was because by dive bombing, I changed my altitude from 5, to 3, during the weapons resolution phase. So, during the targeting phase, he was still able to assign attacks, but during the resolution phase, MM saw the changed altitude, and created a 2 hex deadzone around my fighter.

I think this is really cool, and kind of want it to stay in so your fighters can "hit the deck" to escape an attack, but it might be a bug. Save games attached.

[DeadzonesCancelAttacks.sav.gz](https://github.com/MegaMek/megamek/files/465697/DeadzonesCancelAttacks.sav.gz)
[autosave_2016-09-10_16-42-22.sav.gz](https://github.com/MegaMek/megamek/files/465698/autosave_2016-09-10_16-42-22.sav.gz)

**Comment by VampireSeraphin** [comment]
Saw something similar in a second game tonight where the aero attacking a ground hex while under air to air attack caused the AA shots to vanish. This time the range was longer than the dead zone that should have been created though. 

[autosave_2016-09-10_22-30-20.sav.gz](https://github.com/MegaMek/megamek/files/465737/autosave_2016-09-10_22-30-20.sav.gz)

**Comment by ObviousTech** [comment]
Same game, other side.

Aero X at elevation 3.
Aero Y at elevation 4 nearby.

X can't fire at Y because it's in a dead zone.
Y fires at ground unit Z.  Elevation changes to 3.
X now is able to fire at Y, is no longer in a dead zone.  Shots process correctly.

**Comment by arlith** [comment]
Yes, I believe this is a (probably unintended) consequence of the errata that changed the altitude change from Aero attacks to the beginning of weapons fire instead of the end.  I kind of believe that the reported case is a little disconcerting for the user though.  I'm curious to see if this situation was considered when they made the errata changes...

**Comment by scJazz2** [comment]
I have been hunting high and low for the Errata referencing this topic. Any hints Arlith?

**Comment by VampireSeraphin** [comment]
I kind of like the idea that aerospace fighters can "hit the deck" to avoid air to air attacks by attacking a ground unit or ground hex. It's a neat trick, and I love those, but it's also definitely exploitable, because every time the air unit under attack changes altitude, it seems to cancel the attacks from other fighters.

If it's going to be left in, can we get two things please. First, a change to the LOS cone to show dead zones. Second, a notification in the firing report of "attack canceled because target changed altitude." Just something so that it's more clear to other users whats going on under the hood.

**Comment by arlith** [comment]
TW Erra v4.3, pg 39:

> 2 **Striking** (p. 245)
First paragraph, third sentence
    _A strike attack reduces the attacker’s altitude by one._
Change to:
    _A strike attack reduces the attacker’s altitude by one, applied before any attacks are resolved that_ turn.

**Comment by arlith** [comment]
Yea, I don't really care that it happens.  It's not anything I think we will fix, unless the rules get changed.  Basically, it's working as written.

What I don't like, is that some attacks get canceled because of this, and it's not reported to the user.  I agree that something should be printed, but I think it's going to be a big pain to implement it.

Essentially, the way it works is, the client queues up a list of `AttackAction`s on the client, which are validated.  Then, these attacks are sent to to the server, which stores them and then processes the attacks.  When the server processes the attack, it checks to see if they're validate.  If they aren't valid, it just skips it.  The "checking for validate" bit is a big chunk of code, and right now I don't think there's any mechanism for being able to distinguish when an attack was valid, but then an ASF drops into a dead zone.  Printing a message for this is going to require some awkward code I think.

There's also a timing question.  The rules are vague.  It just says, "applied before any attacks are resolved."  There's an attack declaration phase, and a firing resolution phase.  You could technically apply the altitude lose as soon as the ASF makes the dive/strafe, which would eliminate the issue without sim fire, when the ASF fires first, but doesn't eliminate the problem if the ASF is already shot at, then makes a strike attack.  I interpreted the rules to mean the altitude lose is applied after all attacks are declared, but before firing is resolved (which is how I believe it works now, but I may be wrong).

I dunno, I think in general this errata was made for dubious reasons...

**Comment by VampireSeraphin** [comment]
Yeah, I don't mind when actions fail generally. I mind when the program doesn't report back to you WHY they failed.

**Comment by arlith** [comment]
Yes, I agree with you 100%.  The tricky bit here is that I don't know that the framework supports this reporting without some fairly awkward code.

**Comment by VampireSeraphin** [comment]
As an interim, could you have MM just spit out a "shot canceled, invalid target" when that happens during the firing phase? That would be less work than trying to get a code that sorts all the reasons, and still let the player know that something went awry.

**Comment by rjhancock** [comment]
Closing due to age. Please test with the latest milestone, 0.50.11, to see if the issue still persists. If it does, please re-open.

**Comment by Gribbly1** [comment]
It still happens:

[a2gdeadzone1.sav.gz](https://github.com/user-attachments/files/24727071/a2gdeadzone1.sav.gz)
[a2gdeadzone2.sav.gz](https://github.com/user-attachments/files/24727073/a2gdeadzone2.sav.gz)
[a2gdeadzone3.sav.gz](https://github.com/user-attachments/files/24727072/a2gdeadzone3.sav.gz)

name: test

It's working as intended due to

<img width="395" height="129" alt="Image" src="https://github.com/user-attachments/assets/affac512-e64c-4810-8b05-58f8bdce6b4b" />

Users notice this from time to time. It's dependent on initiative and firing order, at least if and when the air to air firing opportunity appears. One change that has happened at some point in the last couple of years is that the altitude change is now immediately displayed after the striking unit hits 'done' in the firing phase, which makes things clearer half the time.

If the rule was that the altitude loss from striking happened in the end phase it wouldn't be a problem.

**Comment by gsparks3** [comment]
There is actually an easily-missed line (possibly missed even when they added the "applied before any attacks are resolved that turn" to the section shown above in errata) on TW page 242 that resolves this issue:
>Air-to-ground attacks must be declared during the Movement (Aerospace) Phase, including any change in altitude and skill rolls that occur as a result.

I am uncertain how difficult it would be to properly implement this in Megamek. Presumably during movement, as part of moving an aerospace unit, you would have to push a button to indicate intent to either strike or dive-bomb, with the altitude changes taking effect immediately. Then the attack would be made normally during the Firing phase.

**Comment by Gribbly1** [comment]
That would make it possible for a following ASF (that's won initiative) to drop altitude and still fire at a strike aircraft it's pursuing. It would also remove an erroneous opportunity to fire in the firing phase when it's lost initiative.

**Comment by gsparks3** [comment]
Actually, rereading that section, it also notes that an aerospace unit making an air-to-ground attack can't make any air-to-air attacks that turn, so probably all such attacks should indeed be declared during movement (since "this unit can't shoot at me this turn" is a fairly significant piece of information for other aerospace units moving afterwards). At a minimum, declare the intent to make an A2G attack, and how much altitude you lose by doing so (which locks in two out of the four possible attack types).

**Comment by HammerGS** [comment]
Related - https://github.com/MegaMek/megamek/issues/5192
https://github.com/MegaMek/megamek/issues/7620
